### ISSUE TITLE
GPU prediction does not work on Android

### ISSUE THREAD

**Issue by joseph-o3h** [natmlx/natml#50]
I get these errors in logcat when running a prediction with a model that uses the GPU
```
E/tflite: TfLiteGpuDelegate Invoke: GpuDelegate must run on the same thread where it was initialized.
E/tflite: Node number 374 (TfLiteGpuDelegateV2) failed to invoke.
```

This probably means that `Predict` must be called on the same thread that created the GpuDelegate, but with the async `MLEdgeModel.Create` it is not known on which thread it is created.

There should be a non-async variant of this method that can be run on a background thread controlled by the app so that the predictions can also be run on the same thread.

**Comment by olokobayusuf** [comment]
@joseph-o3h can you share the full, unfiltered logs from this? Feel free to shoot me a PM [on Discord](https://www.natml.ai/community) or email with the logs if it contains sensitive info. I ask because this issue doesn't seem to be uniformly reproducible; I suspect that only the OpenGL delegate has this issue, and not the OpenCL delegate (which takes precedence).

In any case for safety, the GPU delegate will be disabled by default on Android, making it opt-in. We'll need to find a better way to handle this than forcing model creation to be synchronous.

**Comment by olokobayusuf** [comment]
@joseph-o3h what version of Android is your test device running? I just found out that on Android S+, apps need to explicitly [request access to vendor libraries](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/java/AndroidManifestGpu.xml) (OpenCL falls into this category because Android doesn't officially support it).  This would explain why TFLite falls back to OpenGL, which then results in this GPU thread error.

**Comment by joseph-o3h** [comment]
Yes it was the OpenGL delegate that was causing this (here is the full log from tflite)
```
W/tflite: NNAPI SL driver did not implement SL_ANeuralNetworksDiagnostic_registerCallbacks!
I/tflite: Replacing 204 node(s) with delegate (TfLiteNnapiDelegate) node, yielding 1 partitions.
E/tflite: Following operations are not supported by GPU delegate:
    DELEGATE TfLiteNnapiDelegate: Operation is not supported.
    No operations will run on the GPU, and all 1 operations will run on the CPU.
I/tflite: Created 0 GPU delegate kernels.
W/tflite: NNAPI SL driver did not implement SL_ANeuralNetworksDiagnostic_registerCallbacks!
I/tflite: Replacing 365 node(s) with delegate (TfLiteNnapiDelegate) node, yielding 5 partitions.
W/tflite: NNAPI SL driver did not implement SL_ANeuralNetworksDiagnostic_registerCallbacks!
W/tflite: NNAPI SL driver did not implement SL_ANeuralNetworksDiagnostic_registerCallbacks!
W/tflite: NNAPI SL driver did not implement SL_ANeuralNetworksDiagnostic_registerCallbacks!
E/tflite: Following operations are not supported by GPU delegate:
    DELEGATE TfLiteNnapiDelegate: Operation is not supported.
    5 operations will run on the GPU, and the remaining 4 operations will run on the CPU.
I/tflite: Replacing 5 node(s) with delegate (TfLiteGpuDelegateV2) node, yielding 3 partitions.
E/tflite: Can not open OpenCL library on this device - dlopen failed: library "libOpenCL.so" not found
E/tflite: Falling back to OpenGL
I/tflite: Created 1 GPU delegate kernels.
```
The device was running Android 13.

**Comment by olokobayusuf** [comment]
Yup relevant lines:
```
E/tflite: Can not open OpenCL library on this device - dlopen failed: library "libOpenCL.so" not found
E/tflite: Falling back to OpenGL
```
The next update will add the `<uses-native-library>` tag that should let the application load `libOpenCL.so`. And we'll remove the GPU from being used by default. NatML Hub will maintain a configuration database for which Android devices can use the GPU, so models fetched from Hub will be able to leverage the GPU when possible.

**Comment by joseph-o3h** [comment]
> NatML Hub will maintain a configuration database for which Android devices can use the GPU, so models fetched from Hub will be able to leverage the GPU when possible.

But what if the model is embedded? In that case can the GPU be enabled by detecting the presence of libOpenCL.so on the device at runtime?

**Comment by olokobayusuf** [comment]
> But what if the model is embedded? In that case can the GPU be enabled by detecting the presence of libOpenCL.so on the device at runtime?

The issue isn't necessarily about `libOpenCL.so` being available at runtime (it's pretty common); it's that on Android S+, apps have to explicitly declare that they will use vendor libraries. I just found out that NatML can't declare this just yet, because the version of Gradle (Android build plugin) that Unity 2022.1 or older uses doesn't support `<uses-native-library>`.

So for now, the simplest workaround you all can do is to detect the version of Android at runtime, and if it's older than Android S, then enable the GPU compute target. In the next update, NatML will default to using the CPU and neural processor (and not the GPU) on Android.

**Comment by olokobayusuf** [comment]
Following up from convo above, NatML now does not use the GPU by default on Android. We'll be able to do so once we can bump the minimum Unity version to Unity 2022.2. See the [NatML 1.1.4 changelog](https://github.com/natmlx/natml/releases/tag/1.1.4).
